spam/02_spam_data_augmentation_tutorial.py

```diff
@@ -140,47 +143,53 @@ def swap_adjectives(x):
 
 
 def get_synonym(word, pos=None):
+    """Get synonym for word given its part-of-speech (pos)."""
     synsets = wn.synsets(word, pos=pos)
-    if not synsets:
-        return None
-    else:
+    if synsets:
         words = [lemma.name() for lemma in synsets[0].lemmas()]
-        return words[0] if words else word
+        if words[0].lower() != word.lower():  # Skip if synonym is same as word.
+            # Multi word synonyms in wordnet use '_' as a separator e.g. reckon_with. Replace it with space.
+            return words[0].replace("_", " ")
+
+
+def replace_token(spacy_doc, idx, replacement):
+    """Replace token in position idx with replacement."""
+    return " ".join([spacy_doc[:idx].text, replacement, spacy_doc[1 + idx :].text])
 
 
 @transformation_function(pre=[spacy])
 def replace_verb_with_synonym(x):
-    words = [token.text for token in x.doc]
-    idxs = [i for i, token in enumerate(x.doc) if token.pos_ == "VERB"]
-    if len(idxs) > 1:
-        to_replace = np.random.choice(idxs[:-1])
-        synonym = get_synonym(words[to_replace], pos="v")
-        if synonym and synonym != words[to_replace]:
-            x.text = " ".join(words[:to_replace] + [synonym] + words[1 + to_replace :])
+    verb_idxs = [i for i, token in enumerate(x.doc) if token.pos_ == "VERB"]
+    if verb_idxs:
+        # Pick random verb idx to replace.
+        idx = np.random.choice(verb_idxs)
+        synonym = get_synonym(x.doc[idx].text, pos="v")
+        if synonym:
+            x.text = replace_token(x.doc, idx, synonym)
             return x
 
 
 @transformation_function(pre=[spacy])
 def replace_noun_with_synonym(x):
-    words = [token.text for token in x.doc]
-    idxs = [i for i, token in enumerate(x.doc) if token.pos_ == "NOUN"]
-    if len(idxs) > 1:
-        to_replace = np.random.choice(idxs[:-1])
-        synonym = get_synonym(words[to_replace], pos="n")
-        if synonym and synonym != words[to_replace]:
-            x.text = " ".join(words[:to_replace] + [synonym] + words[1 + to_replace :])
+    noun_idxs = [i for i, token in enumerate(x.doc) if token.pos_ == "NOUN"]
+    if noun_idxs:
+        # Pick random noun idx to replace.
+        idx = np.random.choice(noun_idxs)
+        synonym = get_synonym(x.doc[idx].text, pos="n")
+        if synonym:
```

Comment:
more comments! e.g. # If there's a valid noun synonym, replace it. Otherwise, return None